File: bin/MadGraph5_aMCatNLO/gridpack_test/BulkGraviton_hh_GF_HH_narrow_M900_gridpack/work/MG5_aMC_v2_4_2/aloha/template_files/txxxxx.cc
#include <complex>
#include <cmath> 
#include "aloha_aux_functions.h"
using namespace std;
#include <iostream>
#include <cstdlib>

void txxxxx(double p[4], double tmass, int nhel, int nst, complex<double> tc[18])
{
	complex<double> ft[6][4], ep[4], em[4], e0[4];
	double pt, pt2, pp, pzpt, emp, sqh, sqs;
	int i, j;
	
	sqh = pow( 0.5, 0.5 );
	sqs = pow( 0.5/3, 0.5 );
	
	pt2 = p[1]*p[1] + p[2]*p[2];
	pp = min( p[0], pow( pt2+p[3]*p[3], 0.5 ) );
	pt = min( pp, pow( pt2, 0.5 ) );
	
	ft[4][0] = complex<double>( p[0]*nst, p[3]*nst );
	ft[5][0] = complex<double>( p[1]*nst, p[2]*nst );
	
	// construct eps+
	if( nhel >= 0 )
	{
		if( pp == 0 )
		{
			ep[0] = complex<double>( 0, 0 );
			ep[1] = complex<double>( -sqh, 0 );
			ep[2] = complex<double>( 0, nst*sqh );
			ep[3] = complex<double>( 0, 0 );
		}
		else
		{
			ep[0] = complex<double>( 0, 0 );
			ep[3] = complex<double>( pt/pp*sqh, 0 );
			
			if( pt != 0 )
			{
				pzpt = p[3]/(pp*pt)*sqh;
				ep[1] = complex<double>( -p[1]*pzpt, -nst*p[2]/pt*sqh );
				ep[2] = complex<double>( -p[2]*pzpt, nst*p[1]/pt*sqh );
			}
			else
			{
				ep[1] = complex<double>( -sqh, 0 );
				ep[2] = complex<double>( 0, nst*Sgn(sqh,p[3]) );
			}
		}
		
	}
	
	// construct eps-
	if( nhel <= 0 )
	{
		if( pp == 0 )
		{
			em[0] = complex<double>( 0, 0 );
			em[1] = complex<double>( sqh, 0 );
			em[2] = complex<double>( 0, nst*sqh );
			em[3] = complex<double>( 0, 0 );
		}
		else
		{
			em[0] = complex<double>( 0, 0 );
			em[3] = complex<double>( -pt/pp*sqh, 0 );
			
			if( pt != 0 )
			{
				pzpt = -p[3]/(pp*pt)*sqh;
				em[1] = complex<double>( -p[1]*pzpt, -nst*p[2]/pt*sqh );
				em[2] = complex<double>( -p[2]*pzpt,  nst*p[1]/pt*sqh );
			}
			else
			{
				em[1] = complex<double>( sqh, 0 );
				em[2] = complex<double>( 0, nst*Sgn(sqh,p[3]) );
			}
		}
	}
	
	// construct eps0
	if( fabs(nhel) <= 1 )
	{
		if( pp == 0 )
		{
			e0[0] = complex<double>( 0, 0 );
			e0[1] = complex<double>( 0, 0 );
			e0[2] = complex<double>( 0, 0 );
			e0[3] = complex<double>( 1, 0 );
		}
		else
		{
			emp = p[0]/(tmass*pp);
			e0[0] = complex<double>( pp/tmass, 0 );
			e0[3] = complex<double>( p[3]*emp, 0 );
			
			if( pt != 0 )
			{
				e0[1] = complex<double>( p[1]*emp, 0 );
				e0[2] = complex<double>( p[2]*emp, 0 );
			}
			else
			{
				e0[1] = complex<double>( 0, 0 );
				e0[2] = complex<double>( 0, 0 );
			}
		}
	}
	
	if( nhel == 2 )
	{
		for( j=0; j<4; j++ )
		{
			for( i=0; i<4; i++ ) ft[i][j] = ep[i]*ep[j];
		}
	}
	else if( nhel == -2 )
	{
		for( j=0; j<4; j++ )
		{
			for( i=0; i<4; i++ ) ft[i][j] = em[i]*em[j];
		}
	}
	else if( tmass == 0 )
	{
		for( j=0; j<4; j++ )
		{
			for( i=0; i<4; i++ ) ft[i][j] = 0;
		}
	}
	else if( tmass != 0 )
	{
		if( nhel == 1 )
		{
			for( j=0; j<4; j++ )
			{
				for( i=0; i<4; i++ ) ft[i][j] = sqh*( ep[i]*e0[j] + e0[i]*ep[j] );
			}
		}
		else if( nhel == 0 )
		{
			for( j=0; j<4; j++ )
			{
				for( i=0; i<4; i++ ) ft[i][j] = sqs*( ep[i]*em[j] + em[i]*ep[j]
					+ 2.0*e0[i]*e0[j] );
			}
		}
		else if( nhel == -1 )
		{
			for( j=0; j<4; j++ )
			{
				for( i=0; i<4; i++ ) ft[i][j] = sqh*( em[i]*e0[j] + e0[i]*em[j] );
			}
		}
		else
		{
			std::cerr << "Invalid helicity in txxxxx.\n";
			std::exit(1);
		}
	}
	
	tc[0] = ft[4][0];
	tc[1] = ft[5][0];
	
	for( j=0; j<4; j++ )
	{
		for( i=0; i<4; i++ ) tc[j*4+i+2] = ft[j][i];
	}
}

File: bin/MadGraph5_aMCatNLO/ewkwmjjjj_5f_LO/ewkwmjjjj_5f_LO_gridpack/work/MG5_aMC_v2_9_13/HELAS/txxxxx.f
      subroutine txxxxx(p,tmass,nhel,nst , tc)
c
c This subroutine computes a TENSOR wavefunction.
c
c input:
c       real    p(0:3)         : four-momentum of tensor boson
c       real    tmass          : mass          of tensor boson
c       integer nhel           : helicity      of tensor boson
c                = -2,-1,0,1,2 : (0 is forbidden if tmass=0.0)
c       integer nst  = -1 or 1 : +1 for final, -1 for initial
c
c output:
c       complex tc(18)         : tensor wavefunction    epsilon^mu^nu(t)
c     
      implicit none
      double precision p(0:3), tmass
      integer nhel, nst
      double complex tc(18)

      double complex ft(6,4), ep(4), em(4), e0(4)
      double precision pt, pt2, pp, pzpt, emp, sqh, sqs
      integer i, j

      double precision rZero, rHalf, rOne, rTwo
      parameter( rZero = 0.0d0, rHalf = 0.5d0 )
      parameter( rOne = 1.0d0, rTwo = 2.0d0 )

      integer stdo
      parameter( stdo = 6 )

      sqh = sqrt(rHalf)
      sqs = sqrt(rHalf/3.d0)

      pt2 = p(1)**2 + p(2)**2
      pp = min(p(0),sqrt(pt2+p(3)**2))
      pt = min(pp,sqrt(pt2))

      ft(5,1) = dcmplx(p(0),p(3))*nst
      ft(6,1) = dcmplx(p(1),p(2))*nst

      if ( nhel.ge.0 ) then 
c construct eps+
         if ( pp.eq.rZero ) then
            ep(1) = dcmplx( rZero )
            ep(2) = dcmplx( -sqh )
            ep(3) = dcmplx( rZero , nst*sqh )
            ep(4) = dcmplx( rZero )
         else
            ep(1) = dcmplx( rZero )
            ep(4) = dcmplx( pt/pp*sqh )
            if ( pt.ne.rZero ) then
               pzpt = p(3)/(pp*pt)*sqh
               ep(2) = dcmplx( -p(1)*pzpt , -nst*p(2)/pt*sqh )
               ep(3) = dcmplx( -p(2)*pzpt ,  nst*p(1)/pt*sqh )
            else
               ep(2) = dcmplx( -sqh )
               ep(3) = dcmplx( rZero , nst*sign(sqh,p(3)) )
            endif
         endif
      end if

      if ( nhel.le.0 ) then 
c construct eps-
         if ( pp.eq.rZero ) then
            em(1) = dcmplx( rZero )
            em(2) = dcmplx( sqh )
            em(3) = dcmplx( rZero , nst*sqh )
            em(4) = dcmplx( rZero )
         else
            em(1) = dcmplx( rZero )
            em(4) = dcmplx( -pt/pp*sqh )
            if ( pt.ne.rZero ) then
               pzpt = -p(3)/(pp*pt)*sqh
               em(2) = dcmplx( -p(1)*pzpt , -nst*p(2)/pt*sqh )
               em(3) = dcmplx( -p(2)*pzpt ,  nst*p(1)/pt*sqh )
            else
               em(2) = dcmplx( sqh )
               em(3) = dcmplx( rZero , nst*sign(sqh,p(3)) )
            endif
         endif
      end if

      if ( abs(nhel).le.1 ) then  
c construct eps0
         if ( pp.eq.rZero ) then
            e0(1) = dcmplx( rZero )
            e0(2) = dcmplx( rZero )
            e0(3) = dcmplx( rZero )
            e0(4) = dcmplx( rOne )
         else
            emp = p(0)/(tmass*pp)
            e0(1) = dcmplx( pp/tmass )
            e0(4) = dcmplx( p(3)*emp )
            if ( pt.ne.rZero ) then
               e0(2) = dcmplx( p(1)*emp )
               e0(3) = dcmplx( p(2)*emp )
            else
               e0(2) = dcmplx( rZero )
               e0(3) = dcmplx( rZero )
            endif
         end if
      end if

      if ( nhel.eq.2 ) then
         do j = 1,4
            do i = 1,4
               ft(i,j) = ep(i)*ep(j)
            end do
         end do
      else if ( nhel.eq.-2 ) then
         do j = 1,4
            do i = 1,4
               ft(i,j) = em(i)*em(j)
            end do
         end do
      else if (tmass.eq.0) then
         do j = 1,4
            do i = 1,4
               ft(i,j) = 0
            end do
         end do
      else if (tmass.ne.0) then
        if  ( nhel.eq.1 ) then
           do j = 1,4
              do i = 1,4
                 ft(i,j) = sqh*( ep(i)*e0(j) + e0(i)*ep(j) )
              end do
           end do
        else if ( nhel.eq.0 ) then
           do j = 1,4
              do i = 1,4
                 ft(i,j) = sqs*( ep(i)*em(j) + em(i)*ep(j)
     &                                + rTwo*e0(i)*e0(j) )
              end do
           end do
        else if ( nhel.eq.-1 ) then
           do j = 1,4
              do i = 1,4
                 ft(i,j) = sqh*( em(i)*e0(j) + e0(i)*em(j) )
              end do
           end do
        else
           write(stdo,*) 'invalid helicity in TXXXXX'
           stop
        end if
      end if

      tc(1) = ft(1,1)
      tc(2) = ft(1,2)
      tc(3) = ft(1,3)
      tc(4) = ft(1,4)
      tc(5) = ft(2,1)
      tc(6) = ft(2,2)
      tc(7) = ft(2,3)
      tc(8) = ft(2,4)
      tc(9) = ft(3,1)
      tc(10) = ft(3,2)
      tc(11) = ft(3,3)
      tc(12) = ft(3,4)
      tc(13) = ft(4,1)
      tc(14) = ft(4,2)
      tc(15) = ft(4,3)
      tc(16) = ft(4,4)
      tc(17) = ft(5,1)
      tc(18) = ft(6,1)

      return
      end
